File: python/nvagentrt/tools.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Tool lifecycle operations.

Provides both manual and managed tool-call workflows.

Functions:
    call(name, args, *, handle=None, attributes=None, data=None, metadata=None, tool_call_id=None)
        Begin a tool call manually. Returns a ``ToolHandle`` that must later
        be passed to ``call_end``. Emits a ``Start`` event. The optional
        ``tool_call_id`` is an external correlation ID propagated to events
        for ATIF trajectory linking.

    call_end(handle, result, *, data=None, metadata=None)
        End a manual tool call. Records the result and emits an ``End`` event.

    execute(name, args, func, *, handle=None, attributes=None, data=None, metadata=None)
        Execute a tool call through the full middleware pipeline (conditional-
        execution guardrails on raw args → request intercepts →
        sanitize-request guardrails → execution intercepts → *func* →
        response intercepts → sanitize-response guardrails). On rejection,
        only a standalone Mark event is emitted and ``GuardrailRejected`` is
        raised. Returns an awaitable of the final result.

    request_intercepts(name, args)
        Run the registered tool request intercept chain on the given arguments.
        Returns the transformed arguments.

    conditional_execution(name, args)
        Run the registered tool conditional execution guardrail chain.
        Raises ``RuntimeError`` if any guardrail rejects.

    response_intercepts(name, result)
        Run the registered tool response intercept chain on the given result.
        Returns the transformed result.

Example::

    import nvagentrt

    # Managed execution (recommended)
    result = await nvagentrt.tools.execute("search", {"q": "hello"}, my_search)

    # Manual lifecycle
    handle = nvagentrt.tools.call("search", {"q": "hello"})
    result = my_search({"q": "hello"})
    nvagentrt.tools.call_end(handle, result)
"""

from nvagentrt._native import (
    nvagentrt_tool_call as call,
)
from nvagentrt._native import (
    nvagentrt_tool_call_end as call_end,
)
from nvagentrt._native import (
    nvagentrt_tool_call_execute as execute,
)
from nvagentrt._native import (
    nvagentrt_tool_conditional_execution as conditional_execution,
)
from nvagentrt._native import (
    nvagentrt_tool_request_intercepts as request_intercepts,
)
from nvagentrt._native import (
    nvagentrt_tool_response_intercepts as response_intercepts,
)

__all__ = ["call", "call_end", "execute", "request_intercepts", "conditional_execution", "response_intercepts"]

File: go/nvagentrt/tools/tools.go
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Package tools provides shorthand access to NVAgentRT tool call operations.
//
// It re-exports the core tool lifecycle functions (ToolCall, ToolCallEnd,
// ToolCallExecute) under shorter names for convenience.
//
// Example usage:
//
//	import "github.com/nvidia/nvagentrt/go/nvagentrt/tools"
//
//	// Execute a tool call with an inline function.
//	result, err := tools.Execute("search", json.RawMessage(`{"q":"hello"}`),
//	    func(args json.RawMessage) (json.RawMessage, error) {
//	        // ... perform the search ...
//	        return json.RawMessage(`{"results":[]}`), nil
//	    },
//	)
package tools

import (
	"encoding/json"

	"github.com/nvidia/nvagentrt/go/nvagentrt"
)

// Call starts a tool call lifecycle and returns a [nvagentrt.ToolHandle],
// emitting a Start event. End the call with [CallEnd]. This is a shorthand for
// [nvagentrt.ToolCall].
func Call(name string, args json.RawMessage, opts ...nvagentrt.ToolCallOption) (*nvagentrt.ToolHandle, error) {
	return nvagentrt.ToolCall(name, args, opts...)
}

// CallEnd completes a tool call that was started with [Call], emitting an End
// event. This is a shorthand for [nvagentrt.ToolCallEnd].
func CallEnd(handle *nvagentrt.ToolHandle, result json.RawMessage, opts ...nvagentrt.ToolCallOption) error {
	return nvagentrt.ToolCallEnd(handle, result, opts...)
}

// Execute runs a complete tool call lifecycle with the full middleware pipeline
// (conditional-execution guardrails, request intercepts, sanitize-request
// guardrails, execution intercepts, fn, response intercepts, sanitize-response
// guardrails) and returns the final result JSON. This is a shorthand for
// [nvagentrt.ToolCallExecute].
func Execute(name string, args json.RawMessage, fn nvagentrt.ToolExecutionFunc, opts ...nvagentrt.ToolCallOption) (json.RawMessage, error) {
	return nvagentrt.ToolCallExecute(name, args, fn, opts...)
}

// RequestIntercepts runs the registered tool request intercept chain on the
// given arguments and returns the transformed arguments. This is a shorthand for
// [nvagentrt.ToolRequestIntercepts].
func RequestIntercepts(name string, args json.RawMessage) (json.RawMessage, error) {
	return nvagentrt.ToolRequestIntercepts(name, args)
}

// ConditionalExecution runs the registered tool conditional execution guardrail
// chain. Returns nil if all guardrails pass, or an error with the rejection
// reason if blocked. This is a shorthand for [nvagentrt.ToolConditionalExecution].
func ConditionalExecution(name string, args json.RawMessage) error {
	return nvagentrt.ToolConditionalExecution(name, args)
}

// ResponseIntercepts runs the registered tool response intercept chain on the
// given result and returns the transformed result. This is a shorthand for
// [nvagentrt.ToolResponseIntercepts].
func ResponseIntercepts(name string, result json.RawMessage) (json.RawMessage, error) {
	return nvagentrt.ToolResponseIntercepts(name, result)
}
